Implement the Cookie Store API get/set/delete functions for Service Workers

RupinMittal opened the pull request:
#### 2ecb71d1b38b5f41d1c2cfd7baf45972e01c76d7
<pre>
Implement the Cookie Store API get/set/delete functions for Service Workers
<a href="https://bugs.webkit.org/show_bug.cgi?id=260737">https://bugs.webkit.org/show_bug.cgi?id=260737</a>

Reviewed by Chris Dumez.

The Cookie Store API should be supported for Service Workers. This patch
adds Service Worker support for the get/set/delete cookie functions.
The Cookie Store is now created using a ScriptExecutionContext rather
the CookieJar where needed.

Since get/set/delete are called through the CookieJar, and CookieJar is
a main thread object, we now have a CookieStore::MainThreadBridge which
is used to hop between the main thread and the service worker thread
when needed. This will also be used for cookie change notifications in
a future patch.

* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieListItem_attributes.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_delete_arguments.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_delete_basic.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_getAll_arguments.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_getAll_multiple.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_getAll_set_basic.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_arguments.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_delete_basic.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_set_across_frames.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_set_basic.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_set_ordering.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_opaque_origin.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_set_arguments.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_special_names.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/encoding.https.any.serviceworker-expected.txt:
* Source/WebCore/Modules/cookie-store/CookieInit.h:
(WebCore::CookieInit::isolatedCopy const):
(WebCore::CookieInit::isolatedCopy):
* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::MainThreadBridge::create):
(WebCore::CookieStore::MainThreadBridge::MainThreadBridge):
(WebCore::CookieStore::MainThreadBridge::ensureOnMainThread):
(WebCore::CookieStore::MainThreadBridge::ensureOnContextThread):
(WebCore::CookieStore::MainThreadBridge::get):
(WebCore::CookieStore::MainThreadBridge::getAll):
(WebCore::CookieStore::MainThreadBridge::set):
(WebCore::CookieStore::create):
(WebCore::CookieStore::CookieStore):
(WebCore::CookieStore::get):
(WebCore::CookieStore::getAll):
(WebCore::CookieStore::set):
(WebCore::CookieStore::cookiesAdded):
(WebCore::CookieStore::cookiesDeleted):
(WebCore::CookieStore::stop):
(WebCore::CookieStore::eventListenersDidChange):
(WebCore::CookieStore::takePromise):
* Source/WebCore/Modules/cookie-store/CookieStore.h:
* Source/WebCore/Modules/cookie-store/CookieStoreGetOptions.h:
(WebCore::CookieStoreGetOptions::isolatedCopy const):
(WebCore::CookieStoreGetOptions::isolatedCopy):
* Source/WebCore/loader/CookieJar.cpp:
(WebCore::CookieJar::sameSiteInfo):
* Source/WebCore/platform/Cookie.h:
(WebCore::Cookie::Cookie):
(WebCore::Cookie::isolatedCopy const):
(WebCore::Cookie::isolatedCopy):
* Source/WebCore/workers/service/ServiceWorkerGlobalScope.cpp:
(WebCore::ServiceWorkerGlobalScope::cookieStore):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::cookiesForDOMAsync):
(WebKit::NetworkConnectionToWebProcess::setCookieFromDOMAsync):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
(WebKit::WebSWContextManagerConnection::installServiceWorker):
* Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
(WebKit::WebCookieJar::getCookiesAsync const):
(WebKit::WebCookieJar::setCookieAsync const):

Canonical link: <a href="https://commits.webkit.org/268055@main">https://commits.webkit.org/268055@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bfa34c879a09a5d7584b2a030a5e67d7662e392e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18503 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18838 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19441 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20357 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17309 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/18699 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22150 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18985 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/19192 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18726 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18902 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16114 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21233 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16125 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16876 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/23335 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17144 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17046 "9 flakes 1 failures") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21231 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17624 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/14940 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16686 "Built successfully") | | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4406 "layout-tests (failure)") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21050 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2278 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17466 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +60,25 @@ namespace WebCore {
-    auto* page = document->page();
-    if (!page)
-        return;
+    if (context->isDocument()) {
```
`is<Document>(context)`.

Your current code is not safe as context might be null.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +60,25 @@ namespace WebCore {
 
-    m_cookieJar = page->cookieJar();
-    m_host = document->url().host().toString();
+        m_cookieJar = page->cookieJar();
```
Why do we still need this?

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +60,25 @@ namespace WebCore {
-    m_cookieJar = page->cookieJar();
-    m_host = document->url().host().toString();
+        m_cookieJar = page->cookieJar();
+        m_host = document->url().host().toString();
```
ditto.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -210,79 +249,98 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
-    if (!cookie.domain.isNull()) {
-        if (cookie.domain.startsWith('.')) {
-            promise->reject(Exception { TypeError, "The domain must not begin with a '.'"_s });
+    auto setCookies = [options = WTFMove(options), promise = WTFMove(promise)](ScriptExecutionContext* context) {
```
How about we take a `ScriptExecutionContext&` since we know it is non-null?

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -210,79 +249,98 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
 
-    m_cookieJar->setCookieAsync(document, document.url(), cookie, WTFMove(completionHandler));
+    if (context->isDocument()) {
+        setCookies(context);
```
`setCookies(*context);

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -210,79 +249,98 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
-            promise->reject(Exception { TypeError, "The domain must not begin with a '.'"_s });
+    auto setCookies = [options = WTFMove(options), promise = WTFMove(promise)](ScriptExecutionContext* context) {
+        auto* document = downcast<Document>(context);
+        if (!document) {
```
Shouldn't be needed.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -210,79 +249,98 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
+        return;
+    }
+
+    downcast<WorkerGlobalScope>(*context).thread().workerLoaderProxy().postTaskToLoader([setCookies = WTFMove(setCookies)](auto& context) mutable {
```
`downcast<WorkerGlobalScope>(*context).thread().workerLoaderProxy().postTaskToLoader(WTFMove(setCookies));`

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -210,79 +241,111 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
-    cookie.name = WTFMove(options.name);
-    cookie.value = WTFMove(options.value);
-    cookie.created = WallTime::now().secondsSinceEpoch().milliseconds();
+    auto setCookies = [options = WTFMove(options), promise = WTFMove(promise)](ScriptExecutionContext& context) {
```
`options = crossThreadCopy(WTFMove(options))` now that multiple threads are involved.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -210,79 +241,111 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
+
+            auto host = document.url().host().toString();
+            if (!host.endsWith(cookie.domain) || (host.length() > cookie.domain.length() && !StringView(host).substring(0, host.length() - cookie.domain.length()).endsWith('.'))) {
+                promise->reject(Exception { TypeError, "The domain must be a part of the current host"_s });
```
This is not OK in the worker case. You're rejecting the promise on the main thread instead of the worker thread.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -210,79 +241,111 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
-    }
+        auto& cookieJar = page->cookieJar();
+        auto completionHandler = [promise = WTFMove(promise)] (bool setSuccessfully) {
+            ALWAYS_LOG_WITH_STREAM(stream << "rupin log - point 7 - in completionHandler");
```
I assume this is getting called on the main thread, even in the worker case, which means you're resolving the promise on the wrong thread even in the success case.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -278,49 +278,33 @@ void WebCookieJar::deleteCookie(const WebCore::Document& document, const URL& ur
     ApplyTrackingPrevention applyTrackingPreventionInNetworkProcess = ApplyTrackingPrevention::No;
 #if ENABLE(TRACKING_PREVENTION)
-    if (shouldBlockCookies(webFrame.get(), document.firstPartyForCookies(), url, applyTrackingPreventionInNetworkProcess)) {
+    if (shouldBlockCookies(nullptr, document.firstPartyForCookies(), url, applyTrackingPreventionInNetworkProcess)) {
```
Note that this is now incorrect in the non service-worker case. We're supposed to pass a WebFrame if we have one.

RupinMittal replied:
Fixed!

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -278,49 +278,33 @@ void WebCookieJar::deleteCookie(const WebCore::Document& document, const URL& ur
-    auto pageID = webFrame->page()->identifier();
 
-    WebProcess::singleton().ensureNetworkProcessConnection().connection().sendWithAsyncReply(Messages::NetworkConnectionToWebProcess::SetCookieFromDOMAsync(document.firstPartyForCookies(), sameSiteInfo, url, frameID, pageID, applyTrackingPreventionInNetworkProcess, shouldRelaxThirdPartyCookieBlocking(webFrame.get()), cookie), WTFMove(completionHandler));
+    WebProcess::singleton().ensureNetworkProcessConnection().connection().sendWithAsyncReply(Messages::NetworkConnectionToWebProcess::SetCookieFromDOMAsync(document.firstPartyForCookies(), sameSiteInfo, url, std::nullopt, std::nullopt, applyTrackingPreventionInNetworkProcess, shouldRelaxThirdPartyCookieBlocking(nullptr), cookie), WTFMove(completionHandler));
```
We're supposed to pass the pageID / frameID if we have them (in the window case).

RupinMittal replied:
Fixed!

youennf commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -103,28 +97,39 @@ void CookieStore::get(CookieStoreGetOptions&& options, Ref<DeferredPromise>&& pr
-    auto completionHandler = [promise = WTFMove(promise)] (std::optional<Vector<Cookie>>&& cookies) {
-        if (!cookies) {
-            promise->reject(TypeError);
+    auto getCookies = [options = WTFMove(options), promise = WTFMove(promise)](ScriptExecutionContext& context) {
```
We most probably do not want to capture the promise here, there is a risk it gets unrefed in main thread.
You can probably use a Vector or a HashMap in CookieStore to store it and retrieve it as part of the completion handler.
Or we should ensure the completion handler to always be destroyed in the creation thread, see how WorkerCacheStorageConnection does it.
One possibility would be for m_cookieJar to be the one handling whether it is used by a service worker or a document.

Hoping to main thread with options probably requires isolate copying options as well.
With auto getCookies = lambda, we had issues in the past where getCookies would be copy constructed without us noticing it, resulting in thread safety issues. It might be best to type it as a Function from the start.

RupinMittal replied:
I have fixed the promise issue by using a HashMap in CookieStore!

youennf commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -103,28 +97,39 @@ void CookieStore::get(CookieStoreGetOptions&& options, Ref<DeferredPromise>&& pr
+            promise->resolve<IDLDictionary<CookieListItem>>(CookieListItem(WTFMove(cookiesVector[0])));
+        };
+
+        cookieJar.getCookiesAsync(document, document.url(), options, WTFMove(completionHandler));
```
I think using the document should work for service workers given document is constructed with the script URL and should have all the parameters of the service worker.
Maybe we will uncover differences and might have to change Page::setupForRemoteWorker if so.

youennf commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -103,28 +97,39 @@ void CookieStore::get(CookieStoreGetOptions&& options, Ref<DeferredPromise>&& pr
+            }
 
-        promise->resolve<IDLDictionary<CookieListItem>>(CookieListItem(WTFMove(cookiesVector[0])));
+            promise->resolve<IDLDictionary<CookieListItem>>(CookieListItem(WTFMove(cookiesVector[0])));
```
We would need to hop back to the worker thread here before resolving the promise.

Also why only the first item? Should we specialise the cookie jar to only return 1 item?

RupinMittal replied:
Fixed the code to hop back. The spec (https://wicg.github.io/cookie-store/#CookieStore-get) specifics that we should get the first item only.

cdumez commented on LayoutTests/imported/w3c/web-platform-tests/cookie-store/change_eventhandler_for_document_cookie.https.window-expected.txt:
```diff
@@ -2,5 +2,5 @@
 PASS CookieStore set/overwrite/delete observed by document.cookie
 FAIL CookieStore agrees with document.cookie on encoding non-ASCII cookies assert_equals: Cookie we wrote using document.cookie in cookie jar expected (string) "DOCUMENT-🍪=🔵" but got (undefined) undefined
-PASS document.cookie agrees with CookieStore on encoding non-ASCII cookies
+FAIL document.cookie agrees with CookieStore on encoding non-ASCII cookies assert_equals: Cookie we wrote using CookieStore in document.cookie expected (string) "DOCUMENT-🍪=🔵" but got (undefined) undefined
```
I don't think this patch should cause new failures?

RupinMittal replied:
Fixed!

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +63,291 @@ namespace WebCore {
 
-    auto* page = document->page();
-    if (!page)
+    auto* context = m_cookieStore->scriptExecutionContext();
```
`RefPtr context`

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +63,291 @@ namespace WebCore {
+    ASSERT(context->isContextThread());
+
+    Ref protectedThis { *this };
+    if (auto* document = dynamicDowncast<Document>(*context)) {
```
`RefPtr document`

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +63,291 @@ namespace WebCore {
+        auto& document = downcast<Document>(context);
+        auto* page = document.page();
+        if (!page) {
+            ensureOnContextThread(context.identifier(), [cookieStore, promiseIdentifier]() {
```
This is not using the right context identifier. This is a no-op, it is dispatching to the main thread unconditionally.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +63,291 @@ namespace WebCore {
+};
+
+CookieStore::MainThreadBridge::MainThreadBridge(CookieStore& cookieStore)
+    : m_cookieStore(cookieStore)
```
We probably need to store the identifier of the cookieStore's context as a data member too, so that we can do ensureOnContextThread() properly

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +63,291 @@ namespace WebCore {
+    downcast<WorkerGlobalScope>(*context).thread().workerLoaderProxy().postTaskToLoader(WTFMove(task));
+}
+
+void CookieStore::MainThreadBridge::ensureOnContextThread(ScriptExecutionContextIdentifier contextIdentifier, Function<void()>&& task)
```
This function shouldn't need to take a contextIdentifier.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +63,291 @@ namespace WebCore {
+        return;
+
+    auto getCookies = [this, cookieStore = Ref { *m_cookieStore }, options = WTFMove(options), promiseIdentifier](ScriptExecutionContext& context) {
+        auto& document = downcast<Document>(context);
```
`Ref document`

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +63,291 @@ namespace WebCore {
+        }
+
+        auto& cookieJar = page->cookieJar();
+        auto completionHandler = [this, contextIdentifier = context.identifier(), cookieStore, promiseIdentifier] (std::optional<Vector<Cookie>>&& cookies) {
```
Capturing `this` without `protectedThis` is likely unsafe here.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +63,291 @@ namespace WebCore {
+    ScriptExecutionContext::ensureOnContextThread(contextIdentifier, WTFMove(task));
+}
+
+void CookieStore::MainThreadBridge::get(CookieStoreGetOptions&& options, int promiseIdentifier)
```
It would be nicer if this function took a `CompletionHandler<void(std::optional<Vector<Cookie>>&&)>` instead of a promiseIdentifier. Same comment applies to other MainThreadBride function. I think promise settling should stay in CookieStore, not in the bridge.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +65,243 @@ namespace WebCore {
+
+    void get(CookieStoreGetOptions&&, CompletionHandler<void(ExceptionOr<Vector<Cookie>>&&)>&&);
+    void getAll(CookieStoreGetOptions&&, RefPtr<SecurityOrigin>, bool globalObjectIsWindowObject, CompletionHandler<void(ExceptionOr<Vector<Cookie>>&&)>&&);
+    void set(CookieInit&& options, CompletionHandler<void(ExceptionOr<void>&&)>&&);
```
I think std::optional<Exception> would be nicer than ExceptionOr<void>.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +65,243 @@ namespace WebCore {
+private:
+    MainThreadBridge(CookieStore&, ScriptExecutionContextIdentifier);
+
+    void ensureOnMainThread(Function<void(ScriptExecutionContext& context)>&&);
```
Can omit `context`.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +65,243 @@ namespace WebCore {
         return;
+    ASSERT(context->isContextThread());
+
+    Ref protectedThis { *this };
```
Shouldn't be needed.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +65,243 @@ namespace WebCore {
+
+void CookieStore::MainThreadBridge::get(CookieStoreGetOptions&& options, CompletionHandler<void(ExceptionOr<Vector<Cookie>>&&)>&& completionHandler)
+{
+    auto getCookies = [this, protectedThis = Ref { *this }, options = WTFMove(options), completionHandler = WTFMove(completionHandler)](ScriptExecutionContext& context) mutable {
```
`crossThreadCopy(WTFMove(options))` since it contains String members and you're sending this to another thread.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +65,243 @@ namespace WebCore {
+
+        auto& cookieJar = page->cookieJar();
+        auto resultHandler = [this, protectedThis = WTFMove(protectedThis), completionHandler = WTFMove(completionHandler)] (std::optional<Vector<Cookie>>&& cookies) mutable {
+            ensureOnContextThread([completionHandler = WTFMove(completionHandler), cookies = WTFMove(cookies)]() mutable {
```
`crossThreadCopy(WTFMove(cookies))` since you're sending them to another thread.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +65,243 @@ namespace WebCore {
+
+void CookieStore::MainThreadBridge::getAll(CookieStoreGetOptions&& options, RefPtr<SecurityOrigin> origin, bool globalObjectIsWindowObject, CompletionHandler<void(ExceptionOr<Vector<Cookie>>&&)>&& completionHandler)
+{
+    auto getAllCookies = [this, protectedThis = Ref { *this }, options = WTFMove(options), origin, globalObjectIsWindowObject, completionHandler = WTFMove(completionHandler)](ScriptExecutionContext& context) mutable {
```
`crossThreadCopy(WTFMove(options))`

cdumez replied:
You cannot send `origin` to another thread like this either.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +65,243 @@ namespace WebCore {
+        }
+
+        auto url = document->url();
+        if (!options.url.isNull()) {
```
Why are these checks in the main thread bridge? It doesn't seem like they need to run on the main thread so it should probably move to CookieStore.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +65,243 @@ namespace WebCore {
+
+void CookieStore::MainThreadBridge::set(CookieInit&& options, CompletionHandler<void(ExceptionOr<void>&&)>&& completionHandler)
+{
+    auto setCookie = [this, protectedThis = Ref { *this }, options = crossThreadCopy(WTFMove(options)), completionHandler = WTFMove(completionHandler)](ScriptExecutionContext& context) mutable {
```
Funny how you property called crossThreadCopy() in only one of your 3 functions.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +65,243 @@ namespace WebCore {
+        cookie.created = WallTime::now().secondsSinceEpoch().milliseconds();
+
+        cookie.domain = options.domain.isNull() ? document->domain() : WTFMove(options.domain);
+        if (!cookie.domain.isNull()) {
```
Seems like most or all of these checks don't need to happen on the main thread and could be moved to CookieStore instead of the MainThreadBridge.

cdumez replied:
The cookie could be entirely constructed on the caller side.

cdumez replied:
The cookie could be entirely constructed on the caller side.

RupinMittal replied:
But these checks require us to have a document. If we do them on the caller side, where will we get the document in the case where the caller side is called on a service worker thread?

cdumez replied:
They don't really require a Document, a ScriptExecutionContext should suffice.

From what I gather, you need the domain and the URL host. The domain is `securityOrigin().domain()` which you can call on ScriptExecutionContext. The host can be retrieved from the securityOrigin() too.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +65,243 @@ namespace WebCore {
+
+CookieStore::CookieStore(ScriptExecutionContext* context)
+    : ActiveDOMObject(context)
+    , m_mainThreadBridge(MainThreadBridge::create(*this, context->identifier()))
```
context may be null since it is a raw pointer so it needs a null check.

RupinMittal replied:
Good point. What should we do if context is null? Is there some error we should throw since we will be unable to get an identifier and thus, unable to construct a mainThreadBridge?

cdumez replied:
Just use a default constructed context identifier (which is an invalid identifier)

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +65,243 @@ namespace WebCore {
+    : ActiveDOMObject(context)
+    , m_mainThreadBridge(MainThreadBridge::create(*this, context->identifier()))
+{
+    nextPromiseIdentifier = 0;
```
Definitely not the right place to do this.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -103,28 +330,29 @@ void CookieStore::get(CookieStoreGetOptions&& options, Ref<DeferredPromise>&& pr
-        promise->reject(SecurityError);
-        return;
-    }
+    m_promises.add(nextPromiseIdentifier, WTFMove(promise));
```
`++nextPromiseIdentifier`

Note that you cannot store 0 as key in the HashMap because of our hash traits.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -103,28 +330,29 @@ void CookieStore::get(CookieStoreGetOptions&& options, Ref<DeferredPromise>&& pr
-        return;
-    }
+    m_promises.add(nextPromiseIdentifier, WTFMove(promise));
+    nextPromiseIdentifier++;
```
Not needed.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -103,28 +330,29 @@ void CookieStore::get(CookieStoreGetOptions&& options, Ref<DeferredPromise>&& pr
+    m_promises.add(nextPromiseIdentifier, WTFMove(promise));
+    nextPromiseIdentifier++;
+
+    CompletionHandler<void(ExceptionOr<Vector<Cookie>>&&)> completionHandler = [cookieStore = Ref { *this }, promiseIdentifier = nextPromiseIdentifier - 1](ExceptionOr<Vector<Cookie>>&& result) {
```
Shouldn't need the  `- 1` here any longer.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -103,28 +330,29 @@ void CookieStore::get(CookieStoreGetOptions&& options, Ref<DeferredPromise>&& pr
+    m_promises.add(nextPromiseIdentifier, WTFMove(promise));
+    nextPromiseIdentifier++;
+
+    CompletionHandler<void(ExceptionOr<Vector<Cookie>>&&)> completionHandler = [cookieStore = Ref { *this }, promiseIdentifier = nextPromiseIdentifier - 1](ExceptionOr<Vector<Cookie>>&& result) {
```
I don't think you'll be able to use CompletionHandler or capture cookieStore like this. The reason is that the worker thread may exit while the code is running on the main thread and ensureOnContextThread() will fail, causing this lambda to get destroyed on the main thread.

I think you'll want to use WTF::Function instead. Also, don't capture the cookieStore and have the MainThreadBridge pass it as parameter to the completion handler (it can do it because it hold a WeakPtr to the cookieStore as data member).

RupinMittal replied:
Oh I see. I can replace CompletionHandler with WTF::Function. What is the difference between the two?

cdumez replied:
A CompletionHandler has to be called and can only be called once. It also must be called on the same thread it was constructed on.

A Function doesn't have any of these restrictions. Can not be called, Can be called more than once and on any thread.

RupinMittal replied:
Ah got it. That makes sense.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -392,18 +564,40 @@ ScriptExecutionContext* CookieStore::scriptExecutionContext() const
 #endif
 }
 
+std::optional<Ref<DeferredPromise>> CookieStore::promise(int promiseIdentifier)
```
You never clear promises so this map will just keep growing forever.

This should be named `takePromise()` and actually *take* the promise from the map.

Also, the return type is wrong, an optional Ref<> is a RefPtr<>.

Also, please use uint64_t to the identifier.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.h:
```diff
@@ -82,9 +86,16 @@ class CookieStore final : public RefCounted<CookieStore>, public EventTarget, pu
+    class MainThreadBridge;
+    Ref<MainThreadBridge> m_mainThreadBridge;
+
+    std::optional<Ref<DeferredPromise>> promise(int promiseIdentifier);
```
Please don't mix data members and member functions.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.h:
```diff
@@ -82,9 +86,16 @@ class CookieStore final : public RefCounted<CookieStore>, public EventTarget, pu
     bool m_hasChangeEventListener { false };
     WeakPtr<CookieJar> m_cookieJar;
     String m_host;
+    int nextPromiseIdentifier;
```
`uint64_t m_nextPromiseIdentifier { 0 };`

3 fixes here:
- type to uint64_t
- `m_` prefix for data members
- Inline initialization

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.h:
```diff
@@ -82,9 +86,16 @@ class CookieStore final : public RefCounted<CookieStore>, public EventTarget, pu
     WeakPtr<CookieJar> m_cookieJar;
     String m_host;
+    int nextPromiseIdentifier;
+    HashMap<int, Ref<DeferredPromise>> m_promises;
```
uint64_t.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -278,49 +280,96 @@ void WebCookieJar::deleteCookie(const WebCore::Document& document, const URL& ur
     }
 #endif
 
     auto sameSiteInfo = CookieJar::sameSiteInfo(document, IsForDOMCookieAccess::Yes);
```
Ideally `CookieJar::sameSiteInfo()` would return the right value for dummy documents associated with service workers instead of post-fixing like you're doing here.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +65,164 @@ namespace WebCore {
+    if (!m_cookieStore)
+        return;
+
+    auto getCookies = [this, protectedThis = Ref { *this }, cookieStore = Ref { *m_cookieStore }, options = crossThreadCopy(WTFMove(options)), completionHandler = WTFMove(completionHandler)](ScriptExecutionContext& context) mutable {
```
As mentioned in my previous patch, we shouldn't be passing the CookieStore to another thread as it is not thread-safe. We should drop `cookieStore = Ref { *m_cookieStore }` here.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +65,164 @@ namespace WebCore {
+        Ref document = downcast<Document>(context);
+        WeakPtr page = document->page();
+        if (!page) {
+            ensureOnContextThread([cookieStore, completionHandler = WTFMove(completionHandler)]() mutable {
```
Capture `protectedThis = WTFMove(protectedThis)` here, and stop capturing the cookieStore.

cdumez replied:
Same comment applies in several places below.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +65,164 @@ namespace WebCore {
+    MainThreadBridge(CookieStore&, ScriptExecutionContextIdentifier);
+
+    void ensureOnMainThread(Function<void(ScriptExecutionContext&)>&&);
+    void ensureOnContextThread(Function<void()>&&);
```
`Function<void(CookieStore*)>&&`

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +65,164 @@ namespace WebCore {
+
+void CookieStore::MainThreadBridge::ensureOnContextThread(Function<void()>&& task)
+{
+    ScriptExecutionContext::ensureOnContextThread(m_contextIdentifier, WTFMove(task));
```
```
ScriptExecutionContext::ensureOnContextThread(m_contextIdentifier, [protectedThis = Ref { *this }, task = WTFMove(task)](auto&) {
    if (RefPtr cookieStore = protectedThis->m_cookieStore.get())
        task(*cookieStore);
});
```

RupinMittal replied:
If we do this, then ensureOnContextThread should look like 
`void ensureOnContextThread(Function<void(CookieStore&)>&&);` instead of 
`void ensureOnContextThread(Function<void(CookieStore*)>&&);` right?

cdumez replied:
Correct, I changed my mind mid-review. I figured there is no point in calling the lambda if the weakPtr is null.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -103,19 +251,18 @@ void CookieStore::get(CookieStoreGetOptions&& options, Ref<DeferredPromise>&& pr
-        return;
-    }
+    m_promises.add(++m_nextPromiseIdentifier, WTFMove(promise));
+    Function<void(Ref<CookieStore>, ExceptionOr<Vector<Cookie>>&&)> completionHandler = [promiseIdentifier = m_nextPromiseIdentifier](Ref<CookieStore> cookieStore, ExceptionOr<Vector<Cookie>>&& result) {
```
`CookieStore&` for the parameter.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.h:
```diff
@@ -82,9 +86,16 @@ class CookieStore final : public RefCounted<CookieStore>, public EventTarget, pu
     WeakPtr<CookieJar> m_cookieJar;
     String m_host;
+    uint64_t m_nextPromiseIdentifier { 0 };
+    HashMap<int, Ref<DeferredPromise>> m_promises;
```
uint64_t for the key.

cdumez commented on Source/WebCore/loader/CookieJar.cpp:
```diff
@@ -63,6 +64,26 @@ IncludeSecureCookies CookieJar::shouldIncludeSecureCookies(const Document& docum
+
+    if (frame->loader().client().isRemoteWorkerFrameLoaderClient()) {
+        auto domain = document.isTopDocument() ? RegistrableDomain(document.securityOrigin().data()) : RegistrableDomain(document.siteForCookies());
+        SameSiteInfo sameSiteInfo { };
```
return { domain.matches(url), false, true };

cdumez commented on Source/WebCore/loader/CookieJar.cpp:
```diff
@@ -63,6 +64,26 @@ IncludeSecureCookies CookieJar::shouldIncludeSecureCookies(const Document& docum
+        return { };
+
+    if (frame->loader().client().isRemoteWorkerFrameLoaderClient()) {
+        auto domain = document.isTopDocument() ? RegistrableDomain(document.securityOrigin().data()) : RegistrableDomain(document.siteForCookies());
```
This check doesn't make sense, if this is a Remote worker, then the document is always a top document. Service workers only have one dummy Document.

cdumez commented on Source/WebCore/loader/CookieJar.cpp:
```diff
@@ -63,6 +64,26 @@ IncludeSecureCookies CookieJar::shouldIncludeSecureCookies(const Document& docum
     return (url.protocolIs("https"_s) && !document.foundMixedContent().contains(SecurityContext::MixedContentType::Active)) ? IncludeSecureCookies::Yes : IncludeSecureCookies::No;
 }
 
+SameSiteInfo CookieJar::sameSiteInfo(const Document& document, const URL& url, IsForDOMCookieAccess isAccessForDOM)
```
I don't think we need the URL parameter? You can probably compare the document.url() and document.firstPartyForCookies() to determine if this service worker is for a first party.

See `Page::setupForRemoteWorker()` for how the Document is set up for remote workers.

cdumez commented on Source/WebCore/loader/CookieJar.cpp:
```diff
@@ -63,6 +64,26 @@ IncludeSecureCookies CookieJar::shouldIncludeSecureCookies(const Document& docum
+    return { };
+}
+
 SameSiteInfo CookieJar::sameSiteInfo(const Document& document, IsForDOMCookieAccess isAccessForDOM)
```
Can we modify this one instead of adding another function?

cdumez commented on Source/WebCore/platform/Cookie.h:
```diff
@@ -114,6 +114,16 @@ struct Cookie {
     }
+
+    Cookie isolatedCopy() const & {
+        Vector<uint16_t> copyOfPorts;
```
This shouldn't be needed, Vector has a copy constructor.

cdumez commented on Source/WebCore/platform/Cookie.h:
```diff
@@ -114,6 +114,16 @@ struct Cookie {
+        for (auto port : ports)
+            copyOfPorts.append(port);
+
+        return { name.isolatedCopy(), value.isolatedCopy(), domain.isolatedCopy(), path.isolatedCopy(), created, expires, httpOnly, secure, session, comment.isolatedCopy(), commentURL.isolatedCopy(), WTFMove(copyOfPorts), sameSite };
```
`WTFMove(copyOfPorts)` -> `ports`

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +65,167 @@ namespace WebCore {
+        return adoptRef(*new MainThreadBridge(cookieStore, contextIdentifier));
+    }
+
+    void get(CookieStoreGetOptions&&, Function<void(Ref<CookieStore>, ExceptionOr<Vector<Cookie>>&&)>&&);
```
CookieStore&

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +65,167 @@ namespace WebCore {
+    }
+
+    void get(CookieStoreGetOptions&&, Function<void(Ref<CookieStore>, ExceptionOr<Vector<Cookie>>&&)>&&);
+    void getAll(CookieStoreGetOptions&&, URL&&, Function<void(Ref<CookieStore>, ExceptionOr<Vector<Cookie>>&&)>&&);
```
CookieStore&

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +65,167 @@ namespace WebCore {
+
+    void get(CookieStoreGetOptions&&, Function<void(Ref<CookieStore>, ExceptionOr<Vector<Cookie>>&&)>&&);
+    void getAll(CookieStoreGetOptions&&, URL&&, Function<void(Ref<CookieStore>, ExceptionOr<Vector<Cookie>>&&)>&&);
+    void set(CookieInit&& options, Cookie&&, Function<void(Ref<CookieStore>, std::optional<Exception>&&)>&&);
```
CookieStore&

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +65,167 @@ namespace WebCore {
+        Ref document = downcast<Document>(context);
+        WeakPtr page = document->page();
+        if (!page) {
+            ensureOnContextThread([protectedThis = WTFMove(protectedThis), completionHandler = WTFMove(completionHandler)](CookieStore& cookieStore) mutable {
```
Unnecessary capture of `protectedThis`.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +65,167 @@ namespace WebCore {
+        }
+
+        auto& cookieJar = page->cookieJar();
+        auto resultHandler = [this, protectedThis = WTFMove(protectedThis), completionHandler = WTFMove(completionHandler)] (std::optional<Vector<Cookie>>&& cookies) mutable {
```
ditto.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +65,167 @@ namespace WebCore {
+        Ref document = downcast<Document>(context);
+        WeakPtr page = document->page();
+        if (!page) {
+            ensureOnContextThread([protectedThis = WTFMove(protectedThis), completionHandler = WTFMove(completionHandler)](CookieStore& cookieStore) mutable {
```
ditto.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +65,167 @@ namespace WebCore {
+        }
+
+        auto& cookieJar = page->cookieJar();
+        auto resultHandler = [this, protectedThis = WTFMove(protectedThis), completionHandler = WTFMove(completionHandler)] (std::optional<Vector<Cookie>>&& cookies) mutable {
```
ditto.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +65,167 @@ namespace WebCore {
+        Ref document = downcast<Document>(context);
+        WeakPtr page = document->page();
+        if (!page) {
+            ensureOnContextThread([protectedThis = WTFMove(protectedThis), completionHandler = WTFMove(completionHandler)](CookieStore& cookieStore) mutable {
```
ditto.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -210,28 +360,24 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
-    auto& document = *downcast<Document>(context);
     // The maximum attribute value size is specified at https://wicg.github.io/cookie-store/#cookie-maximum-attribute-value-size.
     static constexpr auto maximumAttributeValueSize = 1024;
+    auto domain = context->securityOrigin()->domain();
```
Seems like you should reuse `origin` from above.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -210,28 +360,24 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
     // The maximum attribute value size is specified at https://wicg.github.io/cookie-store/#cookie-maximum-attribute-value-size.
     static constexpr auto maximumAttributeValueSize = 1024;
+    auto domain = context->securityOrigin()->domain();
+    auto host = context->securityOrigin()->host();
```
ditto.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -392,18 +558,43 @@ ScriptExecutionContext* CookieStore::scriptExecutionContext() const
 
+RefPtr<DeferredPromise> CookieStore::takePromise(uint64_t promiseIdentifier)
+{
+    auto it = m_promises.find(promiseIdentifier);
```
This should be as easy as `return m_promises.take(promiseIdentifier);`.

If this doesn't work, then use a `RefPtr<>` instead of a `Ref<>` for the value in your HashMap to make it work.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.h:
```diff
@@ -29,6 +29,7 @@
 #include "CookieChangeListener.h"
 #include "CookieJar.h"
 #include "EventTarget.h"
+#include "ScriptExecutionContext.h"
```
Can this be a forward declaration?

cdumez commented on Source/WebCore/loader/CookieJar.cpp:
```diff
@@ -65,6 +67,15 @@ IncludeSecureCookies CookieJar::shouldIncludeSecureCookies(const Document& docum
+    if (!frame)
+        return { };
+
+    if (frame->loader().client().isRemoteWorkerFrameLoaderClient()) {
```
`if (frame && frame->loader().client().isRemoteWorkerFrameLoaderClient()) {`

cdumez commented on Source/WebCore/loader/CookieJar.cpp:
```diff
@@ -65,6 +67,15 @@ IncludeSecureCookies CookieJar::shouldIncludeSecureCookies(const Document& docum
 SameSiteInfo CookieJar::sameSiteInfo(const Document& document, IsForDOMCookieAccess isAccessForDOM)
 {
+    auto frame = document.frame();
+    if (!frame)
```
No need with change suggested below.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -278,49 +280,82 @@ void WebCookieJar::deleteCookie(const WebCore::Document& document, const URL& ur
 {
-    RefPtr webFrame = document.frame() ? WebFrame::fromCoreFrame(*document.frame()) : nullptr;
-    if (!webFrame || !webFrame->page()) {
+    auto frame = document.frame();
```
RefPtr

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -278,49 +280,82 @@ void WebCookieJar::deleteCookie(const WebCore::Document& document, const URL& ur
+            return;
+        }
+    } else {
+        RefPtr webFrame = document.frame() ? WebFrame::fromCoreFrame(*document.frame()) : nullptr;
```
you create a `frame` variable above.

RupinMittal replied:
Do you mean that I don't need this at all and can just use the `frame` variable I already made in place of all the places I use this 'webFrame'? Are we using any properties that webFrame has but frame doesn't?

RupinMittal replied:
Oh I misread the comment. You just mean that I can directly do `WebFrame::fromCoreFrame(*frame)`. Understood.

cdumez replied:
You defined `Ref frame = document.frame()` earlier in this function so use it increase of repeatedly calling `document.frame()`.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -278,49 +280,82 @@ void WebCookieJar::deleteCookie(const WebCore::Document& document, const URL& ur
-    if (shouldBlockCookies(webFrame.get(), document.firstPartyForCookies(), url, applyTrackingPreventionInNetworkProcess)) {
-        completionHandler({ });
-        return;
+    if (frame->loader().client().isRemoteWorkerFrameLoaderClient()) {
```
I don't think we need to special case remote workers in this function. We can just rely on the fact that there is a WebFrame or not.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -278,49 +280,82 @@ void WebCookieJar::deleteCookie(const WebCore::Document& document, const URL& ur
-    auto pageID = webFrame->page()->identifier();
 
-    WebProcess::singleton().ensureNetworkProcessConnection().connection().sendWithAsyncReply(Messages::NetworkConnectionToWebProcess::CookiesForDOMAsync(document.firstPartyForCookies(), sameSiteInfo, url, frameID, pageID, includeSecureCookies, applyTrackingPreventionInNetworkProcess, shouldRelaxThirdPartyCookieBlocking(webFrame.get()), options), WTFMove(completionHandler));
+    WebProcess::singleton().ensureNetworkProcessConnection().connection().sendWithAsyncReply(Messages::NetworkConnectionToWebProcess::CookiesForDOMAsync(document.firstPartyForCookies(), sameSiteInfo, url, frameID, pageID, includeSecureCookies, applyTrackingPreventionInNetworkProcess, shouldRelaxThirdPartyCookieBlocking(nullptr), options), WTFMove(completionHandler));
```
Should keep passing webFrame..

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -278,49 +280,82 @@ void WebCookieJar::deleteCookie(const WebCore::Document& document, const URL& ur
-    if (shouldBlockCookies(webFrame.get(), document.firstPartyForCookies(), url, applyTrackingPreventionInNetworkProcess)) {
-        completionHandler(false);
-        return;
+    if (frame->loader().client().isRemoteWorkerFrameLoaderClient()) {
```
ditto, we should not special case remote workers. It should be a single code path.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -278,49 +280,82 @@ void WebCookieJar::deleteCookie(const WebCore::Document& document, const URL& ur
-    auto pageID = webFrame->page()->identifier();
-
-    WebProcess::singleton().ensureNetworkProcessConnection().connection().sendWithAsyncReply(Messages::NetworkConnectionToWebProcess::SetCookieFromDOMAsync(document.firstPartyForCookies(), sameSiteInfo, url, frameID, pageID, applyTrackingPreventionInNetworkProcess, shouldRelaxThirdPartyCookieBlocking(webFrame.get()), cookie), WTFMove(completionHandler));
+    WebProcess::singleton().ensureNetworkProcessConnection().connection().sendWithAsyncReply(Messages::NetworkConnectionToWebProcess::SetCookieFromDOMAsync(document.firstPartyForCookies(), sameSiteInfo, url, frameID, pageID, applyTrackingPreventionInNetworkProcess, shouldRelaxThirdPartyCookieBlocking(nullptr), cookie), WTFMove(completionHandler));
```
We should keep passing webFrame..

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -278,49 +280,51 @@ void WebCookieJar::deleteCookie(const WebCore::Document& document, const URL& ur
-    if (shouldBlockCookies(webFrame.get(), document.firstPartyForCookies(), url, applyTrackingPreventionInNetworkProcess)) {
-        completionHandler(false);
+    if (shouldBlockCookies(webFrame, document.firstPartyForCookies(), url, applyTrackingPreventionInNetworkProcess)) {
+        completionHandler({ });
```
`{ }` -> `false`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -278,49 +280,51 @@ void WebCookieJar::deleteCookie(const WebCore::Document& document, const URL& ur
     auto sameSiteInfo = CookieJar::sameSiteInfo(document, IsForDOMCookieAccess::Yes);
-    auto frameID = webFrame->frameID();
-    auto pageID = webFrame->page()->identifier();
+    std::optional<WebCore::FrameIdentifier> frameID = webFrame ? std::make_optional(webFrame->frameID()) : std::nullopt;
```
Would `auto` work?

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -278,49 +280,51 @@ void WebCookieJar::deleteCookie(const WebCore::Document& document, const URL& ur
-    auto frameID = webFrame->frameID();
-    auto pageID = webFrame->page()->identifier();
+    std::optional<WebCore::FrameIdentifier> frameID = webFrame ? std::make_optional(webFrame->frameID()) : std::nullopt;
+    std::optional<WebCore::PageIdentifier> pageID = webFrame ? std::make_optional(webFrame->page()->identifier()) : std::nullopt;
```
ditto.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -278,49 +280,51 @@ void WebCookieJar::deleteCookie(const WebCore::Document& document, const URL& ur
         completionHandler(false);
         return;
     }
+    auto webFrame = WebFrame::fromCoreFrame(*frame).get();
```
no `.get()`. We want the `auto` to be a smart pointer for safety.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -278,49 +280,51 @@ void WebCookieJar::deleteCookie(const WebCore::Document& document, const URL& ur
     auto includeSecureCookies = CookieJar::shouldIncludeSecureCookies(document, url);
-    auto frameID = webFrame->frameID();
-    auto pageID = webFrame->page()->identifier();
+    std::optional<WebCore::FrameIdentifier> frameID = webFrame ? std::make_optional(webFrame->frameID()) : std::nullopt;
```
`auto`?

cdumez commented on Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
```diff
@@ -278,49 +280,51 @@ void WebCookieJar::deleteCookie(const WebCore::Document& document, const URL& ur
         completionHandler({ });
         return;
     }
+    auto webFrame = WebFrame::fromCoreFrame(*frame).get();
```
no `.get()`

cdumez commented on Source/WebCore/loader/CookieJar.cpp:
```diff
@@ -43,6 +44,7 @@
 #include <optional>
 #include <wtf/CompletionHandler.h>
 #include <wtf/SystemTracing.h>
+#include <wtf/ThreadSafeRefCounted.h>
```
Surprised that this is needed.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -392,18 +558,37 @@ ScriptExecutionContext* CookieStore::scriptExecutionContext() const
+    if (hadChangeEventListener == m_hasChangeEventListener)
+        return;
+
+    auto* document = downcast<Document>(scriptExecutionContext());
```
RefPtr

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -366,11 +519,24 @@ const char* CookieStore::activeDOMObjectName() const
+    if (!m_hasChangeEventListener)
+        return;
+
+    auto* document = downcast<Document>(scriptExecutionContext());
```
RefPtr

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -275,14 +421,19 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
-        if (!setSuccessfully)
-            promise->reject(TypeError);
+    m_promises.add(++m_nextPromiseIdentifier, WTFMove(promise));
+    Function<void(CookieStore&, std::optional<Exception>&&)> completionHandler = [promiseIdentifier = m_nextPromiseIdentifier](CookieStore& cookieStore, std::optional<Exception>&& result) {
```
can we use `auto`?

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -151,39 +301,39 @@ void CookieStore::getAll(CookieStoreGetOptions&& options, Ref<DeferredPromise>&&
-        if (!cookies) {
-            promise->reject(TypeError);
+    m_promises.add(++m_nextPromiseIdentifier, WTFMove(promise));
+    Function<void(CookieStore&, ExceptionOr<Vector<Cookie>>&&)> completionHandler = [promiseIdentifier = m_nextPromiseIdentifier](CookieStore& cookieStore, ExceptionOr<Vector<Cookie>>&& result) {
```
Can we use `auto`?

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -103,19 +254,18 @@ void CookieStore::get(CookieStoreGetOptions&& options, Ref<DeferredPromise>&& pr
-        return;
-    }
+    m_promises.add(++m_nextPromiseIdentifier, WTFMove(promise));
+    Function<void(CookieStore&, ExceptionOr<Vector<Cookie>>&&)> completionHandler = [promiseIdentifier = m_nextPromiseIdentifier](CookieStore& cookieStore, ExceptionOr<Vector<Cookie>>&& result) {
```
can we use `auto`?

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +65,167 @@ namespace WebCore {
+        }
+
+        auto& cookieJar = page->cookieJar();
+        auto resultHandler = [this, completionHandler = WTFMove(completionHandler)] (bool setSuccessfully) mutable {
```
This is capturing `this` without `protectedThis`, which is unsafe.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +65,167 @@ namespace WebCore {
+        }
+
+        auto& cookieJar = page->cookieJar();
+        auto resultHandler = [this, completionHandler = WTFMove(completionHandler)] (std::optional<Vector<Cookie>>&& cookies) mutable {
```
Capturing `this` without `protectedThis` isn't safe.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +65,167 @@ namespace WebCore {
+void CookieStore::MainThreadBridge::getAll(CookieStoreGetOptions&& options, URL&& url, Function<void(CookieStore&, ExceptionOr<Vector<Cookie>>&&)>&& completionHandler)
+{
+    ASSERT(m_cookieStore);
+    if (!m_cookieStore)
```
Can we drop these if checks in all 3 functions? The assertion should suffice given that they are called by the cookie store.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +65,167 @@ namespace WebCore {
+        }
+
+        auto& cookieJar = page->cookieJar();
+        auto resultHandler = [this, completionHandler = WTFMove(completionHandler)] (std::optional<Vector<Cookie>>&& cookies) mutable {
```
Capturing `this` without `protectedThis` isn't safe.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +65,167 @@ namespace WebCore {
 {
-    if (!document)
+    ASSERT(m_cookieStore);
+    if (!m_cookieStore)
```
I believe we can drop this if check.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +65,167 @@ namespace WebCore {
+
+CookieStore::MainThreadBridge::MainThreadBridge(CookieStore& cookieStore, ScriptExecutionContextIdentifier contextIdentifier)
+    : m_cookieStore(cookieStore)
+    , m_contextIdentifier(contextIdentifier)
```
Note that we could get the context identifier from the `cookieStore` directly, we don't need 2 constructor parameters here.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +65,167 @@ namespace WebCore {
+    void set(CookieInit&& options, Cookie&&, Function<void(CookieStore&, std::optional<Exception>&&)>&&);
+
+private:
+    MainThreadBridge(CookieStore&, ScriptExecutionContextIdentifier);
```
`explicit` once you drop the second parameter.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -56,25 +65,159 @@ namespace WebCore {
+
+CookieStore::MainThreadBridge::MainThreadBridge(CookieStore& cookieStore)
+    : m_cookieStore(cookieStore)
+    , m_contextIdentifier(cookieStore.scriptExecutionContext()->identifier())
```
Missing null check for the context again..

RupinMittal replied:
Ah sorry. Fixed!

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -103,19 +246,18 @@ void CookieStore::get(CookieStoreGetOptions&& options, Ref<DeferredPromise>&& pr
             return;
         }
 
-        auto& cookiesVector = *cookies;
```
We don't need to `releaseReturnValue()` here. You can just do this like we used to:
```
auto& cookies = result.returnValue();
```

Also renaming cookiesVector to cookies would be nice now that there is no longer a naming conflict.

RupinMittal replied:
Should I also change `.releaseException' to just `.exception()`?

cdumez replied:
`.releaseException()` since reject() takes in an `Exception` so you want an r-value reference to call the Exception's move constructor instead of doing a copy.

RupinMittal replied:
If I have the cookies vector as a reference, then I'm not able to construct the `CookieListItem` below by doing `WTFMove(cookies[0])`

RupinMittal replied:
Keeping releaseReturnValue for now.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -151,39 +293,39 @@ void CookieStore::getAll(CookieStoreGetOptions&& options, Ref<DeferredPromise>&&
         }
 
-        promise->resolve<IDLSequence<IDLDictionary<CookieListItem>>>(WTF::map(WTFMove(*cookies), [](auto&& cookie) {
+        auto cookiesVector = result.releaseReturnValue();
```
Same comment here.

RupinMittal replied:
If I have the cookies vector as a reference then I won't able to pass `WTFMove(cookies)` to the `map` function below.

RupinMittal replied:
Keeping releaseReturnValue for now.

cdumez reviewed (approved):
Ok, I think it looks good, assuming EWS is happy.

RupinMittal committed:
Implement the Cookie Store API get/set/delete functions for Service Workers
https://bugs.webkit.org/show_bug.cgi?id=260737

Reviewed by Chris Dumez.

The Cookie Store API should be supported for Service Workers. This patch
adds Service Worker support for the get/set/delete cookie functions.
The Cookie Store is now created using a ScriptExecutionContext rather
the CookieJar where needed.

Since get/set/delete are called through the CookieJar, and CookieJar is
a main thread object, we now have a CookieStore::MainThreadBridge which
is used to hop between the main thread and the service worker thread
when needed. This will also be used for cookie change notifications in
a future patch.

* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieListItem_attributes.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_delete_arguments.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_delete_basic.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_getAll_arguments.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_getAll_multiple.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_getAll_set_basic.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_arguments.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_delete_basic.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_set_across_frames.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_set_basic.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_set_ordering.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_opaque_origin.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_set_arguments.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_special_names.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/encoding.https.any.serviceworker-expected.txt:
* Source/WebCore/Modules/cookie-store/CookieInit.h:
(WebCore::CookieInit::isolatedCopy const):
(WebCore::CookieInit::isolatedCopy):
* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::MainThreadBridge::create):
(WebCore::CookieStore::MainThreadBridge::MainThreadBridge):
(WebCore::CookieStore::MainThreadBridge::ensureOnMainThread):
(WebCore::CookieStore::MainThreadBridge::ensureOnContextThread):
(WebCore::CookieStore::MainThreadBridge::get):
(WebCore::CookieStore::MainThreadBridge::getAll):
(WebCore::CookieStore::MainThreadBridge::set):
(WebCore::CookieStore::create):
(WebCore::CookieStore::CookieStore):
(WebCore::CookieStore::get):
(WebCore::CookieStore::getAll):
(WebCore::CookieStore::set):
(WebCore::CookieStore::cookiesAdded):
(WebCore::CookieStore::cookiesDeleted):
(WebCore::CookieStore::stop):
(WebCore::CookieStore::eventListenersDidChange):
(WebCore::CookieStore::takePromise):
* Source/WebCore/Modules/cookie-store/CookieStore.h:
* Source/WebCore/Modules/cookie-store/CookieStoreGetOptions.h:
(WebCore::CookieStoreGetOptions::isolatedCopy const):
(WebCore::CookieStoreGetOptions::isolatedCopy):
* Source/WebCore/loader/CookieJar.cpp:
(WebCore::CookieJar::sameSiteInfo):
* Source/WebCore/platform/Cookie.h:
(WebCore::Cookie::Cookie):
(WebCore::Cookie::isolatedCopy const):
(WebCore::Cookie::isolatedCopy):
* Source/WebCore/workers/service/ServiceWorkerGlobalScope.cpp:
(WebCore::ServiceWorkerGlobalScope::cookieStore):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::cookiesForDOMAsync):
(WebKit::NetworkConnectionToWebProcess::setCookieFromDOMAsync):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
(WebKit::WebSWContextManagerConnection::installServiceWorker):
* Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
(WebKit::WebCookieJar::getCookiesAsync const):
(WebKit::WebCookieJar::setCookieAsync const):

Canonical link: https://commits.webkit.org/268055@main
